internal/expr/parser.go

```diff
@@ -6,6 +6,79 @@ import (
 	"strings"
 )
 
+// A QueryPart represents a section of a parsed SQL statement, which forms
+// a complete query when processed together with its surrounding parts, in
+// their correct order.
+type queryPart interface {
+	// String returns the part's representation for debugging purposes.
+	String() string
+
+	// marker method
+	part()
+}
+
+// typeName stores a Go type and a member of it.
+type typeName struct {
+	name, member string
+}
+
+func (tn typeName) String() string {
+	return tn.name + "." + tn.member
+}
+
+// columnName stores a SQL column and optionally its table.
+type columnName struct {
+	table, name string
+}
+
+func (cn columnName) String() string {
+	if cn.table == "" {
+		return cn.name
+	}
+	return cn.table + "." + cn.name
+}
+
+// inputPart represents a named parameter that will be sent to the database
+// while performing the query.
+type inputPart struct {
+	sourceType typeName
+	isSlice    bool
+	raw        string
+}
+
+func (p *inputPart) String() string {
+	return fmt.Sprintf("Input[%+v]", p.sourceType)
+}
+
+func (p *inputPart) part() {}
+
+// outputPart represents a named target output variable in the SQL expression,
+// as well as the source table and column where it will be read from.
+type outputPart struct {
+	preparedColumns []columnName
```

Comment:
This might not be important but I would like to store this in a different place, rather than at the Parser output. Before thing change, the parts contained all the information about the parsed state and only that, not carrying extra state. Now, we need to fill the `preparedColumns` attribute before creating the SQL with no indication of it in the type system.

Reply:
Thats very fair, I've changed it to be stored in the PreparedExpr itself.